apps/server/src/shared/repo/user/user-do.repo.ts

```diff
@@ -0,0 +1,97 @@
+	get entityName(): EntityName<User> {
+		return User;
+	}
+
+	getConstructor(): { new (I): User } {
+		return User;
+	}
+
+	async findById(id: EntityId, populate = false): Promise<UserDO> {
+		const userEntity: User = await this._em.findOneOrFail(this.entityName, id as FilterQuery<User>);
+
+		if (populate) {
+			await this._em.populate(userEntity, ['roles', 'school.systems', 'school.schoolYear']);
+			await this.populateRoles(userEntity.roles.getItems());
+		}
+
+		return this.mapEntityToDO(userEntity);
+	}
+
+	async findByExternalIdOrFail(externalId: string, systemId: string): Promise<UserDO> {
+		const userEntitys: User[] = await this._em.find(User, { externalId }, { populate: ['school.systems'] });
+		const userEntity: User | undefined = userEntitys.find((user: User): boolean => {
+			const { systems } = user.school;
+			return systems && !!systems.getItems().find((system: System): boolean => system.id === systemId);
+		});
+
+		if (!userEntity) {
+			throw new NotFoundError(`User entity with externalId: ${externalId} and systemId: ${systemId} not found.`);
+		}
+
+		return this.mapEntityToDO(userEntity);
+	}
+
+	private async populateRoles(roles: Role[]): Promise<void> {
+		for (let i = 0; i < roles.length; i += 1) {
+			const role = roles[i];
+			if (!role.roles.isInitialized(true)) {
+				// eslint-disable-next-line no-await-in-loop
+				await this._em.populate(role, ['roles']);
+				// eslint-disable-next-line no-await-in-loop
+				await this.populateRoles(role.roles.getItems());
+			}
+		}
+	}
+
+	protected mapEntityToDO(entity: User): UserDO {
+		const user: UserDO = new UserDO({
+			id: entity.id,
+			createdAt: entity.createdAt,
+			updatedAt: entity.updatedAt,
+			email: entity.email,
+			firstName: entity.firstName,
+			lastName: entity.lastName,
+			roleIds: [],
+			schoolId: entity.school.id,
+			ldapDn: entity.ldapDn,
+			externalId: entity.externalId,
+			importHash: entity.importHash,
+			firstNameSearchValues: entity.firstNameSearchValues,
+			lastNameSearchValues: entity.lastNameSearchValues,
+			emailSearchValues: entity.emailSearchValues,
+			language: entity.language,
+			forcePasswordChange: entity.forcePasswordChange,
+			preferences: entity.preferences,
+		});
+
+		if (entity.roles.isInitialized(true)) {
+			user.roleIds = entity.roles.getItems().map((role: Role) => role.id);
+		}
+
+		return user;
+	}
+
+	protected mapDOToEntity(entityDO: UserDO): EntityProperties<IUserProperties> {
+		return {
+			id: entityDO.id,
+			email: entityDO.email,
+			firstName: entityDO.firstName,
+			lastName: entityDO.lastName,
+			school: Reference.createFromPK(School, entityDO.schoolId),
+			roles: entityDO.roleIds.map((roleId) => Reference.createFromPK(Role, roleId)),
+			ldapDn: entityDO.ldapDn,
+			externalId: entityDO.externalId,
+			language: entityDO.language,
+			forcePasswordChange: entityDO.forcePasswordChange,
+			preferences: entityDO.preferences,
+		};
+	}
+}
```

Comment:
This code duplication is currently deliberate, since not all UCs have moved to this new Repo format and we need to keep compatibility. Once we are sure that all UCs are moved (And all Repos are ported as well) then the old ones can be removed.

Reply:
I understand, but that's too big a switch for all repos. I think it should just be our BaseRepo with two abstract methods:
- public mapEntityToDO(entity: T | T [], do: P): P []
- public mapDOToEntity(do: P | P[]) T | T[]

These methods are generic and you can pass different DOs. But I'd like to discuss that in the arc meeting.

Reply:
can stay that way, as this is in line with current archetectural design.